File: app/src/main/java/hexlet/code/games/ProgressionGame.java
package hexlet.code.games;

import java.util.Random;
import java.util.StringJoiner;

public final class ProgressionGame extends Game {

    private static final int PROGRESSION_LENGTH = 10;

    private static int[] mageProgression(int start, int step) {
        var progression = new int[PROGRESSION_LENGTH];

        for (var i = 0; i < progression.length; i++) {
            progression[i] = start + i * step;
        }

        return progression;
    }

    @Override
    public String getRule() {
        return "What number is missing in the progression?";
    }

    @Override
    public GameRound getNewRound() {
        var random = new Random();
        var start = random.nextInt(1, 11);
        var step = random.nextInt(1, 6);
        var hiddenIdx = random.nextInt(PROGRESSION_LENGTH);

        var progression = mageProgression(start, step);
        var rightAnswer = Integer.toString(progression[hiddenIdx]);

        return new GameRound(
                prepareTask(progression, hiddenIdx),
                rightAnswer
        );
    }

    private String prepareTask(int[] progression, int hiddenIdx) {
        var task = new StringJoiner(" ");

        for (var i = 0; i < progression.length; i++) {
            if (i == hiddenIdx) {
                task.add("..");
            } else {
                var current = progression[i];
                task.add(Integer.toString(current));
            }
        }

        return task.toString();
    }
}

File: app/src/main/kotlin/hexlet/code/ProgressionGame.kt
package hexlet.code

class ProgressionGame : Game() {
    override fun rule(): String {
        return "What number is missing in the progression?"
    }

    override fun newRound(): GameRound {
        @Suppress("MagicNumber")
        val start = (0..10).random()

        @Suppress("MagicNumber")
        val step = (1..5).random()

        val hiddenElementIndex = (0..<PROGRESSION_LENGTH).random()
        val progression = makeProgression(start, step)

        val rightAnswer = progression[hiddenElementIndex].toString()
        val question = question(progression, hiddenElementIndex)

        return GameRound(question, rightAnswer)
    }

    private fun question(progression: IntArray, hiddenElementIndex: Int): String {
        val progressionString = progression
            .mapIndexed { idx, el -> if (idx == hiddenElementIndex) ".." else el.toString() }
            .joinToString(" ")

        return "Question: $progressionString"
    }

    private fun makeProgression(start: Int, step: Int): IntArray {
        return IntArray(PROGRESSION_LENGTH) { start + it * step }
    }

    companion object {
        private const val PROGRESSION_LENGTH = 10
    }
}
